File: internal/webhook/webhook.go
// Package webhook sends HTTP POST notifications on crawl events.
package webhook

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/micelio/micelio/internal/crawler"
)

// Payload is the JSON body sent to the webhook URL.
type Payload struct {
	Schedule   *Schedule `json:"schedule,omitempty"`
	Event      string    `json:"event"`
	URL        string    `json:"url"`
	Timestamp  string    `json:"timestamp"`
	Duration   string    `json:"duration"`
	OutputFile string    `json:"outputFile"`
	HTMLReport string    `json:"htmlReport,omitempty"`
	Pages      int       `json:"pages"`
	Errors     int       `json:"errors"`
}

// Schedule holds cron context when the crawl was triggered by a schedule.
type Schedule struct {
	Cron      string `json:"cron"`
	NextRun   string `json:"nextRun"`
	RunNumber int    `json:"runNumber"`
}

// Options configures the webhook target.
type Options struct {
	Headers        map[string]string
	URL            string
	SkipValidation bool // skip SSRF validation (for testing with localhost servers)
}

const (
	maxAttempts = 2
	retryDelay  = 2 * time.Second
	sendTimeout = 10 * time.Second
)

// ValidateURL checks if a webhook URL is safe to dispatch to.
// Returns an error if the URL uses a non-HTTP scheme or targets a private/loopback address.
func ValidateURL(rawURL string) error {
	if !strings.HasPrefix(rawURL, "http://") && !strings.HasPrefix(rawURL, "https://") {
		return fmt.Errorf("webhook URL must use http:// or https:// scheme")
	}
	if crawler.IsPrivateURL(rawURL) {
		return fmt.Errorf("webhook URL targets a private/loopback address")
	}
	return nil
}

// Send dispatches a webhook POST with retry. Returns true on success.
// Logs warnings on error but never returns an error.
func Send(opts Options, payload Payload) bool {
	if !opts.SkipValidation {
		if err := ValidateURL(opts.URL); err != nil {
			log.Printf("[webhook] blocked: %v (%s)", err, RedactURL(opts.URL))
			return false
		}
	}

	body, err := json.Marshal(payload)
	if err != nil {
		log.Printf("[webhook] marshal error: %v", err)
		return false
	}

	var client *http.Client
	if opts.SkipValidation {
		client = &http.Client{Timeout: sendTimeout}
	} else {
		client = crawler.SafeClientFollowRedirects(sendTimeout)
	}

	for attempt := 0; attempt < maxAttempts; attempt++ {
		req, err := http.NewRequest(http.MethodPost, opts.URL, bytes.NewReader(body))
		if err != nil {
			log.Printf("[webhook] request error: %v", err)
			return false
		}
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("User-Agent", "Micelio/1.0 Scheduler")
		for k, v := range opts.Headers {
			req.Header.Set(k, v)
		}

		resp, err := client.Do(req)
		if err != nil {
			log.Printf("[webhook] attempt %d failed: %v", attempt+1, err)
		} else {
			resp.Body.Close()
			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
				return true
			}
			log.Printf("[webhook] attempt %d: HTTP %d %s", attempt+1, resp.StatusCode, resp.Status)
		}

		if attempt == 0 {
			time.Sleep(retryDelay)
		}
	}

	log.Printf("[webhook] failed to deliver notification to %s", RedactURL(opts.URL))
	return false
}

// RedactURL strips path and query (may contain tokens) for safe logging.
func RedactURL(rawURL string) string {
	parsed, err := url.Parse(rawURL)
	if err != nil || parsed.Host == "" {
		return rawURL
	}
	return fmt.Sprintf("%s://%s/...", parsed.Scheme, parsed.Host)
}

File: src/webhook.ts
/**
 * Webhook notification sender for scheduled crawl completions.
 */

export interface WebhookPayload {
  event: 'crawl_complete' | 'crawl_failed';
  url: string;
  timestamp: string;
  duration: string;
  pages: number;
  errors: number;
  outputFile: string;
  htmlReport?: string;
  schedule?: {
    cron: string;
    runNumber: number;
    nextRun: string;
  };
}

export interface WebhookOptions {
  url: string;
  headers?: Record<string, string>;
}

/**
 * Send a webhook notification. Retries once on failure.
 * Logs warnings on error but never throws.
 */
export async function sendWebhook(
  options: WebhookOptions,
  payload: WebhookPayload,
): Promise<boolean> {
  const { url, headers = {} } = options;

  const body = JSON.stringify(payload);
  const requestHeaders: Record<string, string> = {
    'Content-Type': 'application/json',
    'User-Agent': 'Micelio/1.0 Scheduler',
    ...headers,
  };

  for (let attempt = 0; attempt < 2; attempt++) {
    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: requestHeaders,
        body,
        signal: AbortSignal.timeout(10_000),
      });

      if (response.ok) {
        return true;
      }

      process.stderr.write(
        `  [webhook] Attempt ${attempt + 1}: HTTP ${response.status} ${response.statusText}\n`,
      );
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      process.stderr.write(`  [webhook] Attempt ${attempt + 1} failed: ${message}\n`);
    }

    // Wait 2 seconds before retry
    if (attempt === 0) {
      await new Promise((r) => setTimeout(r, 2000));
    }
  }

  // Redact URL path/query for logging (may contain tokens)
  let safeUrl = url;
  try {
    const parsed = new URL(url);
    safeUrl = `${parsed.protocol}//${parsed.host}/...`;
  } catch { /* use original */ }
  process.stderr.write(`  [webhook] Failed to deliver notification to ${safeUrl}\n`);
  return false;
}
